Source/Dafny/Compiler-js.cs

```diff
@@ -1042,9 +1042,51 @@ protected class ClassWriter : IClassWriter {
     // ----- Statements -------------------------------------------------------------
 
     protected override void EmitPrintStmt(TargetWriter wr, Expression arg) {
-      wr.Write("process.stdout.write(_dafny.toString(");
-      TrExpr(arg, wr, false);
-      wr.WriteLine("));");
+      bool isString = arg.Type.AsCollectionType != null &&
+                      arg.Type.AsCollectionType.AsSeqType != null &&
+                      arg.Type.AsCollectionType.AsSeqType.Arg.IsCharType;
+      bool isStringLiteral = arg is StringLiteralExpr;
+      bool isGeneric = arg.Type.AsCollectionType != null &&
+                       arg.Type.AsCollectionType.AsSeqType != null &&
+                       arg.Type.AsCollectionType.AsSeqType.Arg.IsTypeParameter;
+      if (isString && isStringLiteral) {
+        // process.stdout.write(_dafny.toString(x));
+        wr.Write("process.stdout.write(_dafny.toString(");
+        TrExpr(arg, wr, false);
+        wr.WriteLine("));");
+      } else if (isString && !isStringLiteral) {
+        // try { process.stdout.write(_dafny.toString((x).join(""))); } catch (_error) { process.stdout.write(_dafny.toString(x)); }
+        wr.Write("try { process.stdout.write(_dafny.toString((");
+        TrExpr(arg, wr, false);
+        wr.WriteLine(").join(\"\")));");
+        wr.Write(" } catch (_error) { process.stdout.write(_dafny.toString(");
+        TrExpr(arg, wr, false);
+        wr.WriteLine(")); }");
+      } else if (!isString && !isGeneric) {
+        // process.stdout.write(_dafny.toString(x));
+        wr.Write("process.stdout.write(_dafny.toString(");
+        TrExpr(arg, wr, false);
+        wr.WriteLine("));");
+      } else { // isGeneric
+        // try { process.stdout.write(_dafny.toString(((x) instanceof Array && typeof((x)[0]) == \"string\") ? (x).join("") : (x))); } catch (_error) { process.stdout.write(_dafny.toString(x)); }
```

Comment:
Simplifying my suggestion for L1058 further and combining it with this line, why not just do all of this inside `_dafny.toString`?

Reply:
Because we need information from dafny types to distinguish various cases. For example, from dafny we know when we have an empty seq<char>, but within JS we cannot distinguish that from other empty sequences.  It is all a bit messy. Plus only within dafny do we know that the sequence has homogeneous elements -- not at all guaranteed in JS.